Refactor lightbox code to fix buggy implementation of the image-list.

hashirsarwar opened the pull request:
Previously, `lightbox.open()` was responsible for retrieving the image data from the DOM, saving it in `asset_map` and finally displaying the image using that data. This implementation wasn't correct for image list at bottom of the lightbox because the `image` parameter passed to lightbox.open() could contain more than one instances of the image that had to be opened.

This commit refactors the lightbox implementation to fix it's behaviour for the bottom image-list. The changes are as follows:

- Seperated the image data parsing logic from the `lightbox.open()` and moved it into parse_image_data().

- Metadata of all the images is stored in the `asset_map` when an image is opened and the image-list is rendered at the bottom.

- Whenever an image-list element is clicked, the image is directly accessed from `asset_map` rather than relying on the original image element still being findable in the DOM.

Fixes #14152.

hashirsarwar commented on static/js/lightbox.js:
```diff
@@ -76,84 +79,14 @@ function display_video(payload) {
-        asset_map.set($source, payload);
-    }
+exports.open = function ($image) {
+    const $preview_src = $image.attr("src");
```
Using `$preview_src` as the key for `asset_map` instead of `$source` in order to avoid extra conditional statements here.

timabbott commented:
@hashirsarwar thanks for working on this.  Unfortunately, your changes are very hard to read, because your first commit does like 5 things at once.  Can you split the first commit into something more readable?  As I described in the issue, the commit extracting the `parse_image_data` function should be a pure code move, with no real code changes in that commit.  And then there should be a series of commits doing the other changes.

Check out the Zulip commit guidelines for more details: https://zulip.readthedocs.io/en/latest/contributing/version-control.html

hashirsarwar commented:
@timabbott kindly review this once again. :)

timabbott commented on static/js/lightbox.js:
```diff
@@ -89,13 +90,8 @@ exports.open = function (image, options) {
     const payload = asset_map.get($preview_src);
-    if (payload === undefined) {
-        payload = exports.parse_image_data($image)
-    }
```
I think removing this breaks the `show_from_selected_message` used by the `v` hotkey, since that calls the lightbox without parsing it explicitly.

I think the right version of this commit should just add a parse call in `render_lightbox_list_images`, without the other changes.  It's only in rendering that list that we generate new HTML objects that could not be parsed normally.

timabbott commented:
@hashirsarwar huge thanks for splitting this series!  The new version is much easier to understand, and in fact I noticed a bug I would have missed reading the previous version (see above).

timabbott commented:
I merged the first commit, since that looked good and will be responsible for most future merge conflicts this PR experiences :).

For the second commit, can you extend the commit message to mention that that for videos the `src` URL should always uniquely determine the video ID, so we're not losing anything with the change in how videos handle things?

timabbott commented:
I had to merge 31e576543169f2bd8a64ed615a4068a65ea23c7e because the first commit failed tests individually; please in the future run the linter and node tests for each commit when doing frontend work.  It's very helpful for our managing complexity by merging individual commits.

Thanks @hashirsarwar!

hashirsarwar commented:
> I had to merge 31e5765 because the first commit failed tests individually; please in the future run the linter and node tests for each commit when doing frontend work. It's very helpful for our managing complexity by merging individual commits.

@timabbott thanks, I've noted it. Rest of the commits are ready for your review. :slightly_smiling_face:

hashirsarwar commented:
@timabbott sorry for missing that; I have fixed it now.

Moreover, I've detected another bug in the current implementation. When you switch between the images from the `image-list` quickly, the console gets swamped with the exceptions stating,

> Uncaught TypeError: Cannot read property 'clientWidth' of null.

Below is a screenshot while inspecting chat.zulip.org to demonstrate this:
![Screenshot from 2020-04-04 15-15-11](https://user-images.githubusercontent.com/45683359/78425017-0edb3480-768b-11ea-8b7f-52a52802ba35.png)
I've added a commit to resolve this as well.

timabbott commented:
I merged this as the series ending with c2ba3e20b77cf71bf35e6d80e31735dfe1138800, after several changes:

* Re-added the commit 90eda9d9fed65cee2df4d94e488ffa4ff31bb2df that I had pushed to this PR to fix a bug and it appears you accidentally removed when you force-pushed.  In the future, when I mention I've pushed changes to your PR, please fetch the changes before doing further work so we don't risk losing my work.  (The commit 95869974b39cf526f215a61b45a0d1d9da43c050 is wrong without my commit).
* Added a comment in 9ffed7afdc807f0cdffaee064117a96eb50bd624 explaining why we're doing that logic there, since it's an important detail that could easily be missed in a future refactor.
* Reordered the commits to fix independent issues before those fixes are needed.
* Squashed your final fixup commit into the commit introducing the bug, after adding a comment explaining the issue.  See 9a6d83e731c80dce35c85884241b5755a9af15c1.  And I think it'd be really valuable if you could try doing the follow-up mentioned in the TODO comment there.

Thanks for cleaning this up @hashirsarwar!  I'm optimistic that this will remove an exception we've had for years.
